### ISSUE TITLE
Inconsistent getOrderTotal return value

### ISSUE THREAD

**Issue by christian-acceseo** [PrestaShop/PrestaShop#37627]
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [x] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

`$cart->getOrderTotal(type: Cart::BOTH)` and `$cart->getOrderTotal(type: Cart::BOTH_WITHOUT_SHIPPING)` should return the same value for orders with free shipping. But under certain circumstances using `Cart::BOTH_WITHOUT_SHIPPING` will provide a result lower than expected

![Image](https://github.com/user-attachments/assets/6e548ce4-0d54-4b9b-a105-bcfea19676bf)

Diving deeper seems to reveal the issue steaming from the call to `Cart::getTotalCalculationCartRules` from `Cart::getOrderTotal`. 
The truthy branch (Cart::BOTH or Cart::ONLY_DISCOUNTS) in `getTotalCalculationCartRules`'s  main if statement gets all cart rules from the database ordered with `ORDER by cr.priority ASC, cr.gift_product DESC`.
That means gift cart rule will apply first, followed by the 50% off rule.

The falsy branch (Cart::BOTH_WITHOUT_SHIPPING) in getTotalCalculationCartRules`'s  main if statement creates the cart rules array _manually_ in a different order. In this example, the 50% off applies first, followed by the gift rule.

 

### Expected behavior

In the example above both calls to `getOrderTotal` should return 50. Probably achievable by using a predictable sorting order in cart rules returned by `Cart::getTotalCalculationCartRules`.

I have tried this and seems to work fine but I am unable to think of possible side effects in other areas.

### Steps to reproduce

1. Configure free shipping.
2. Create a product without taxes (in order to make numbers easier to deal with), $100.
3. Create a product without taxes (in order to make numbers easier to deal with), $10.
4. Create a discount that applies 50% for selected products. Select both.
5. Create a free gift discount selecting the $10 product.
6. Add the $100 to the cart and apply both discounts.

For the screenshot I injected the following code in `CartController:initContent`, right at the top.
```
        dump([
            'Cart::BOTH' => $this->context->cart->getOrderTotal(type: Cart::BOTH),
            'Cart::BOTH_WITHOUT_SHIPPING' => $this->context->cart->getOrderTotal(type: Cart::BOTH_WITHOUT_SHIPPING),
            'Cart::ONLY_PRODUCTS' => $this->context->cart->getOrderTotal(type: Cart::ONLY_PRODUCTS),
            'Cart::ONLY_DISCOUNTS' => $this->context->cart->getOrderTotal(type: Cart::ONLY_DISCOUNTS),
        ]);
```

### PrestaShop version(s) where the bug happened

8.2.0 and 1.7.7.4

### PHP version(s) where the bug happened

8.1.30 and 7.2.34

**Comment by paulnoelcholot** [comment]
Hello @christian-acceseo 

I try to reproduce your issue but as you can see, I can't : ![Image](https://github.com/user-attachments/assets/ea137409-7d17-4984-8411-c47973e5031e)

Did I miss something?

Thanks!

**Comment by christian-acceseo** [comment]
Hi!

The FO is fine, but if you execute `$cart->getOrderTotal(type: Cart::BOTH_WITHOUT_SHIPPING)` with that cart in your context you should get 45€ instead of 50 which is wrong.

That's due to the fact that cart rules are applied in a different order (_order_ as in sorting things in an array, not a customer order).

**Comment by Codencode** [comment]
Yes, the issue lies in the fact that in the else case, the order of the CartRules is different.

https://github.com/PrestaShop/PrestaShop/blob/d310ebadd685def5b117315ace30c590bcd7c731/classes/Cart.php#L2351-L2370

Perhaps we should always use `$cartRules = $this->getCartRules(CartRule::FILTER_ACTION_ALL, false);` and remove the unnecessary CartRules in the else, but I think the code would become heavy.

**Comment by Codencode** [comment]
An example of code could be the following:
```php
    protected function getTotalCalculationCartRules($type, $withShipping)
    {
        if ($withShipping || $type == Cart::ONLY_DISCOUNTS) {
            $cartRules = $this->getCartRules(CartRule::FILTER_ACTION_ALL, false);
        } else {
            // We retrieve all the CartRules to obtain the same "ordering."
            $allCartRules = $this->getCartRules(CartRule::FILTER_ACTION_ALL, false);

            // We retrieve the CartRules with filter = CartRule::FILTER_ACTION_REDUCTION 
            // and the CartRules with filter = CartRule::FILTER_ACTION_GIFT and merge them into a single array
            $cartRules = array_merge(
                $this->getCartRules(CartRule::FILTER_ACTION_REDUCTION, false), 
                $this->getCartRules(CartRule::FILTER_ACTION_GIFT, false)
            );

            // We extract the IDs of the CartRules
            $cartRulesId = [];
            foreach ($cartRules as $cartRule) {
                $cartRulesId[] = (int) $cartRule['id_cart_rule'];
            }

            // We iterate through all the CartRules, removing those that shouldn't be there.
            foreach ($allCartRules as $index => $cartRule) {
                if (!in_array((int) $cartRule['id_cart_rule'], $cartRulesId)) {
                    unset($allCartRules[$index]);
                }
            }

            $cartRules = $allCartRules;
        }

        return $cartRules;
    }
```

It is just an example and should be improved.
This way, I believe it would solve the issue, but this solution requires adding a call to the `$this->getCartRules()` method.

**Comment by paulnoelcholot** [comment]
Hello,

Thanks for reporting this issue!
Ping @PrestaShop/prestashop-core-developers : Could someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by boo-code** [comment]
Reproduced on `9.2.x`. One 100.00 product in the cart, a gift rule on a 10.00 product and a 50 percent rule, both at priority 1, shipping free:

| | rule order handed to the Calculator | total |
| --- | --- | --- |
| `Cart::BOTH` | gift, then 50 percent | 50.00 |
| `Cart::BOTH_WITHOUT_SHIPPING` | 50 percent, then gift | 45.00 |

`ONLY_SHIPPING` returned 0.00, so the two must agree. Same numbers as the report.

The diagnosis in the issue is right: `getCartRules(FILTER_ACTION_ALL)` orders by `priority ASC, gift_product DESC`, while the merge of a `FILTER_ACTION_REDUCTION` result with a `FILTER_ACTION_GIFT` one puts every reduction ahead of every gift whatever the priority.

### The proposed fix does not terminate

Building the subset from `FILTER_ACTION_ALL` and filtering the unwanted rules out - the shape suggested above - hangs as soon as the cart holds a free-shipping rule.

`getCartRules()` passes its `$filter` down to `CartRule::getContextualValue()`. With `FILTER_ACTION_ALL` the free-shipping branch at `classes/CartRule.php:1523` calls `getOrderTotal(..., Cart::ONLY_SHIPPING)`, and the shipping cost is evaluated against `getOrderTotal(true, Cart::BOTH_WITHOUT_SHIPPING)` at `classes/Cart.php:4903`, which re-enters `getTotalCalculationCartRules()`.

Measured: with that version a cart holding a free-shipping rule did not return within 120 seconds; the same cart on unmodified code finished in seconds, and the same version on a cart with no free-shipping rule returned normally. The two narrower filters are what keeps the shipping value from being computed on this path - `classes/Cart.php:2304` already says "dont process free shipping to avoid calculation loop".

So the merge has to stay and only the order can be corrected. `Cart::getOrderedCartRulesIds()` already answers "in what order does this cart apply its rules", computes no contextual value, is cached and invalidated next to `getCartRules()` at `classes/Cart.php:1453-1456`, and is what `CartRule::getContextualValue()` itself uses at `classes/CartRule.php:1495`. Reordering the merged array against it gives `BOTH_WITHOUT_SHIPPING` the same relative order as `BOTH`, and being a sort it cannot change the set. Checked on a cart carrying four rules - reduction only, gift only, free shipping only, and free shipping plus a reduction: `BOTH_WITHOUT_SHIPPING` still excludes the shipping-only rule and still includes the one that also carries a reduction, now in `BOTH`'s relative order.

### A second divergence, under the discount feature flag

With `discount` enabled the totals still differ after that fix, for an independent reason. `getCartRules()` orders its query by `priority ASC, gift_product DESC` before running `sortCartRulesByPriority()`; `getOrderedCartRulesIds()` runs the same sorter over a query with no `ORDER BY`. `DiscountPriority::compareDiscounts()` ends on `strcmp()` of `date_add`, so two rules that tie on type, on the priority field and on the creation second compare equal, and `usort` - stable since PHP 8 - keeps whatever order each query returned. Measured with the flag on: the two methods report opposite orders for the same pair of rules, and `BOTH` and `BOTH_WITHOUT_SHIPPING` are 50.00 and 45.00 again. Adding the same `ORDER BY` to the second query made both report the same order and both totals 50.00.

That one sits inside the discount rework rather than in the legacy path, so I have left it out of the fix rather than changing the new system's ordering unasked.
